docs/docs/building-a-site-with-authentication.md

```diff
@@ -2,33 +2,84 @@
+
+### Setting up client-only routes
+
+With Gatsby, you are able to create restricted areas in your app using [client-only routes](/docs/building-apps-with-gatsby/#client-only-routes).
+
+Because static HTML files generated by Gatsby sit on a file server, you cannot programmatically control access to those files on the server from Gatsby (for example: a user could guess or type in a URL and navigate straight to the page). As the [section from the Adding App and Website functionality](/docs/adding-app-and-website-functionality/#client-only-routes) overview page demonstrates, client-only routes can be created to route a user between pages using a React-based router, as opposed to navigating between different static HTML files on a server. Gatsby is different from a traditional Single Page Application in
+
+Taking advantage of this client-side routing gives the control back to you to protect or customize your routes. Using the [`@reach/router` library](https://reach.tech/router/), which comes installed with Gatsby, you can setup a router on a page and control which component will be loaded when a certain route is called, and check for the existence of a variable like authentication state before serving the content.
+
+<!-- prettier-ignore -->
+```jsx
+<Router>
+  {isAuthenticated ? <PrivateRoute /> : <Login />}
+</Router>
+```
+
+More specific code examples for this pattern are outlined in the [Client-only Routes & User Authentication guide](/docs/client-only-routes-and-user-authentication/#implementing-client-only-routes).
+
+### Protecting code from accessing browser globals during build
+
+Global objects that are accessible in the browser like `localStorage` aren't available while a Gatsby site is building, because the build runs in a Node.js environment.
+
+Some third party services might try and access `localStorage` or the `window` object with internal methods, to keep those snippets from breaking the build, those invocations should be wrapped in checks or `useEffect` hooks to verify that the code is running in the browser and is skipped during the build process:
+
+```javascript
+import app from "firebase/app"
+
+...
+
+if (typeof window !== 'undefined') { // highlight-line
+  app.initializeApp(config)
+} // highlight-line
+```
+
+More information on build related errors is available in the guide on [debugging HTML builds](/docs/debugging-html-builds/).
 
 ## Real-world example: Gatsby store
 
-The [Gatsby store](https://github.com/gatsbyjs/store.gatsbyjs.org) uses this
-approach when handling a private route:
+The [Gatsby store](https://github.com/gatsbyjs/store.gatsbyjs.org) is a live application built with Gatsby that implements authentication using Auth0.
+
+[Util functions](https://github.com/gatsbyjs/store.gatsbyjs.org/blob/master/src/utils/auth.js) in the repo make use of Auth0's APIs while checking that [some code only runs in the browser](https://github.com/gatsbyjs/store.gatsbyjs.org/blob/master/src/utils/auth.js#L3).
```

Comment:
A bit confused by this sentence. Are the util functions making sure the code is wrapped properly? Or is it Auth0?

Reply:
Just made some updates, does it make more sense now?